first pass at changing QueueProcessor/Worker/Routee relationship

nsauro opened the pull request:
This PR is the beginning of changing the relationship between Worker, Routee and QueueProcessor.

Previously, the QueueProcessor only created/destroyed the Worker, and the Worker would be responsible for managing its Routee.  It would need to manage creating and recreating the Routee if it failed.  This led to a lot of complexity in the Worker, surrounding failure/retry states, as well as with orchestration with the Queue and QueueProcessor.

In the updates, the following changes have been made:
- QueueProcessor now has the only reference to the Backend.  When a new Worker needs to be created, the QP will create a new Routee from this Backend, and then use this Routee to create the Worker.
- The Worker will not attempt to rebind or resolve any failures with its Routee.  Now, when the Routee dies, the Worker will Unregister itself from the Queue, and then stop itself.  The QP will then remove this Worker from the Pool.  _we might want to change the ordering of this, so that once this event happens(unregistration), the QP immediately removes the Worker from the pool_.  However, we can think about that later.

The Worker also has some state changes
  - WaitingForWork - This is the initial state where it waits for work, 
  - Working - When the Worker has received Work and is waiting on it to complete
  - Unregistering - When a Worker is waiting to be told that it is safely Unregistered, and can stop itself
  - WaitingToTerminate - When the worker knows that after its Work is complete, to stop itself
  - UnregisteringOutstanding - When the Worker is waiting for both unregistration and work to complete.

Here is a state diagram illustrating these states
![workerstates 1](https://cloud.githubusercontent.com/assets/2262111/16745459/eb872f8c-4783-11e6-9280-348f369b1f9b.png)

I've also taken the time to completely redo the WorkerSpec, and break that up into several separate Specs, one for each state.  This was to ensure that we have very thorough(and hopefully complete!) coverage of the Worker.  I added a lot of new tests.

I commented out a failing test in QueueSpec..not sure why that is failing,and I'm still investigating.

Also, I still need to write some tests for QueueProcessor(which I'm going to rename to WorkerManager, since that fits the description a bit better for what it does).  I need to write tests for all of its new behavior.

Note: this fixes #83 and #84

mwalden commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,170 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
 
-      case work: Work ⇒ sendWorkToDelegatee(work, 0)
+    //ignore these, since all we care about is the Work completing one way or another
+    case Hold(x)             ⇒
```
Are we keeping these here to show the other possible states or can we replace this with _?

nsauro replied:
Wasn't sure if I was going to keep them or not.  I left them while I was doing the Work, just to help me kind of track what was being ignored and not in each state.  While I do like them being there, as they at least show what's actively being ignored, they also do introduce clutter..so I don't have strong feelings either way I suppose.

mwalden replied:
I mean, it helps me personally know what the other possible states are and what to ignore. Just not sure how others feel.

nsauro commented:
One other thing which I forgot to add, is that now, when a Worker completes some Work, it sends a message back to itself `WorkFinished` so that it can then process that event how it wants in its current state.  In the previous implementation, there was some hoop jumping to have the code that handled this event be able to respond in a way consistent with each state it was being executed from.  Hopefully, by sending a complete message to itself, this alleviates that complexity a bit.

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,170 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-        case Right(result) ⇒
-          outstanding.success(result)
-          workFinished(isRetiring)
+  //onRouteeFailure is what gets called if while waiting for a Routee response, the Routee dies.
```
I had to pass in this `onRouteeFailure` function as when I first tried this all out, this code was capturing all of the `Terminated(routee)` stuff, and was handling as it if it was the response from the Routee.  In order to fix that, and _try_ to keep it straightforward, I have this function handling that event, and then calling the functin that each state passes in.  Currently, these are only state transitions.

kailuowang replied:
how about we rename it to `handleRouteeResponse`?

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,170 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+  private def retryOrAbandon(outstanding: Outstanding, error: Any): Unit = {
     outstanding.cancel()
-    //why do we fail if there is a delayBeforeNextWork? Is this because of the subsequent 'sendWorkToDelegatee' call?
+    //why do we fail if there is a delayBeforeNextWork? Is this because of the subsequent 'sendWorkToRoutee' call?
```
This comment is still a valid question :)

kailuowang replied:
mainly because if there is hold, you shouldn't send work to routee, and it's might be too late to try again after the delay. So fail immediately might be the safest option.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -33,14 +34,13 @@ trait QueueProcessor extends Actor with ActorLogging with MessageScheduler {
 
   def receive: Receive = {
-    val workers = (1 to settings.startingPoolSize).map(createWorker).toSet
+    (1 to settings.startingPoolSize).foreach(x ⇒ createRoutee())
```
nit: could be `foreach(_ =>`

nsauro commented:
👊 bump 👊 

I added a QueueProcessorSpec to test all of the new behavior in QueueProcessor.  While doing this, I also made the following changes:

-introduced `WorkerFactory`.  This replaces the function that created Worker Props that was in QueueProcessor.  I did this so that I could inject a different Worker creation function in the tests which created TestProbes.
- Removed `retireQueue` from the `Shutdown` command, as it seemed that was only set to `false` when the QueueProcessor was handling a `Terminated(queue)` message.  I just split the code accordingly.

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+    case RouteeCreated(routee) ⇒
+      workerPool = workerPool + createWorker(routee)
+
+    case RouteeFailed ⇒ //?
```
Right now, the Future which failed to create the Routee is doing logging.  Anything else we want to do here?

kailuowang replied:
if sending and receiving the `RouteeFailed` message doesn't do anything I would get rid of it. If you want to keep it I would move the logging here.  So we know where to put additional logic in the future when creating routee fails.

nsauro replied:
Yea, makes sense.  I can make tht change

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+    case Terminated(worker) if workerPool.contains(worker) ⇒
+      removeWorker(worker)
+
+    //if workers drop below minimum, we have a problem.
```
I still am not sure the best way to handle when a bunch of Workers fail, and fall beneath the minimumWorker threshold.

kailuowang replied:
what's a concern we simply try replenish the worker if the worker pool is below the minWorker threshold and it's not retiring?

nsauro replied:
I will add that, but my concern is really around what happens if we get a continuous stream of failures, due to some network outage, say.  What should the behavior be?

nsauro replied:
We might want to do an exponential back off for failures

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -190,7 +181,17 @@ case class QueueProcessorWithCircuitBreaker(
     }
   }
+}
+
```
new WorkerFactory trait, used in QueueProcessor, and defaults to the DefaultWorkerFactory

kailuowang replied:
how about make it package private? Don't want to give people to false impression that they can override this.

nsauro commented on core/src/test/scala/kanaloa/reactive/dispatcher/queue/QueueProcessorSpec.scala:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    //TODO: not sure why this is broken.
+    "force shutdown if timeout" ignore withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
```
No clue why this is failing.  It just hangs, and I still need to figure out what is going on.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
-      context become monitoring(newHistory)(pool)
-      onWorkError(newHistory, pool)
-    }
+  def workError(): Unit = {
```
now that it's a class scope method, it might be better to rename it to recordWorkError() to be more clear?

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+      val toPoolSize = Math.max(settings.minPoolSize, Math.min(settings.maxPoolSize, newPoolSize))
+      val diff = toPoolSize - workerPool.size
+      if (diff > 0)
+        (1 to diff).foreach { x ⇒
```
nit: `foreach { _ =>`

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
-      log.info(s"Queue Processor is shutdown $withMessage")
-      reportTo.foreach(_ ! ShutdownSuccessfully)
-      context stop self
+  private def createRoutee(): Unit = {
```
it might be better to call it retrieveRoutee, this might gives the false impression that it creates a new Routee every time called. 
And the message should probably be `RouteeRetrieved` as well.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+      workerPool = workerPool + createWorker(routee)
+
+    case RouteeFailed(ex) ⇒
+      log.error(ex, "Failed to create Routee")
```
as suggested by the other comment, the message should probably be "Failed to retrieve Routee"

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
-      context stop self
+  private def createRoutee(): Unit = {
+    import context.dispatcher //do we want to pass this in?
+    backend(this.context).onComplete {
```
not sure why you used `this.`  here.

nsauro commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,153 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+    case Terminated(worker) if workerPool.contains(worker) ⇒
+      removeWorker(worker)
+      if (workerPool.size < settings.minPoolSize) {
+        retrieveRoutee() //kick off the creation of a new Worker
```
now trying to add new Worker if we breach minimum.  I also added a test in QueueProcessorSpec

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    //If there is no work left, or if the Queue dies, the Worker stops as well
+    case NoWorkLeft                   ⇒ finish()
+    case Terminated(`queue`)          ⇒ finish()
+
```
nit: `case NoWorkLeft | Terminated(`queue`) =>`

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+
+    //if the Routee dies or the Worker is told to Retire, it needs to Unregister from the Queue before terminating
+    case Terminated(r) if r == routee ⇒ becomeUnregistering()
+    case Worker.Retire                ⇒ becomeUnregistering()
```
nit: `case Worker.Retire | Terminated(`rourtee`) =>`

kailuowang commented on core/src/test/scala/kanaloa/reactive/dispatcher/IntegrationTests.scala:
```diff
@@ -259,7 +259,7 @@ class AutoScalingWithPullingIntegration extends IntegrationSpec {
       import system.dispatcher
 
-      fishForMessage(duration * 2) {
+      fishForMessage(duration * 3) {
```
I am not sure we should've changed this. It might be an environmental issue. If it's not then the system became slower, which we should find out why. can you try change it back and test one more time?

nsauro commented:
Ok, I'm going to update this branch with all the changes that were just put into master, so after I fix all that, I can circle back to this

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-      }
-      finish()
+    //if there is no work left, or if the Queue dies, the Actor must wait for the Work to finish before terminating
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
```
again I would combine the two cases.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-    case Hold(period)        ⇒ delayBeforeNextWork = Some(period)
+  //in this state we are we waiting for 2 things to happen, Unregistration and Work completing
+  //the Worker will shift its state based on which one happens first
+  def unregisteringOutstanding(outstanding: Outstanding): Receive = routeeResponse(outstanding, becomeUnregistering) orElse {
```
Is it going to send unregister msg to queue again if routee failed in this state?  it might be okay since it's a idempotent operation.

nsauro replied:
yes, if it is in this state, it will send an `unregister` message to the Queue, which shouldn't have any effects.  If need be, I could change the transition here to just 'waitingToTerminate' ?

nsauro replied:
Actually, not sure how safe that it is.  I really wanted to make sure that Workers were unregistered before they terminate, just to avoid any potential work loss.  So, I don't think that one extra unregister message would cause any issues.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-  def whileWaiting(currentStatus: WorkerStatus): Receive = {
-    case NoWorkLeft          ⇒ finish()
+  //in this state, we have told the Queue to Unregister this Worker, so we are waiting for an acknowledgement
+  def unregistering: Receive = {
```
how about rename it `unregisteringWithoutOutstanding`

nsauro replied:
What if we changed these names to both `unregisteringIdle` and `unregisteringBusy`, in the spirit of brevity

kailuowang replied:
sounds good to me.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    case Hold(x)              ⇒
+    case Retire               ⇒
+    case Terminated(`routee`) ⇒
+    case NoWorkLeft           ⇒
```
how about we combine the 4 cases.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    case Hold(x)             ⇒
+    case Retire              ⇒
+    case Terminated(`queue`) ⇒
+    case NoWorkLeft          ⇒
```
how about we combine the 4 cases.

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    //ignore these
+    case Hold(x)             ⇒
+    case Retire              ⇒
+    case NoWorkLeft          ⇒
```
you know I am going to say here 
`case Hold(_) | Retire | NoWorkLeft`

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala:
```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-      (resultChecker orElse ({
+    //Either Unregistration completed, or the Queue died, in either way, we just need to wait for Work to finish
+    case Unregistered        ⇒ context become waitingToTerminate(outstanding)
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
```
combine

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -13,158 +10,153 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
   }
 
+  private def createWorker(routee: ActorRef): WorkerRef = {
+    val workerName = s"worker-${queue.path.name}-${System.nanoTime()}"
```
remember I refactored to use a var worker counter for naming in the other PR?

kailuowang commented on core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala:
```diff
@@ -208,20 +210,24 @@ object QueueProcessor {
   case class RunningStatus(pool: WorkerPool)
   case object ShuttingDown
-  case class Shutdown(reportBackTo: Option[ActorRef] = None, timeout: FiniteDuration = 3.minutes, retireQueue: Boolean = true)
+  private[queue] case class RouteeCreated(routee: ActorRef)
```
rename to `RouteeRetrieved`?

kailuowang commented:
I am not sure how I feel about breaking WorkSpec into different files. I do feel that if someone want to check the tests against worker he would expect to see them one file. I am not against breaking into different classes though, that helps with test-only command. WDYT? 
We are very close now, I think, only some minor comments left.

nsauro commented:
Hrm..I kind of prefer the multi file approach, only because I feel like it improves readability.  I feel like it is easier to digest a couple of smaller test files which are clearly dedicated to testing a single thing, vs having one very large file with everything in it.  The names are pretty clear.  We could put them into their own directory, if you think that will improve discoverability.

kailuowang commented:
It would be awkward if we move into a directory but with a different package name. So I am not sure if it's valid option. 
Let's discuss the multi-file organization. I rarely took this approach before so maybe it's worthwhile to ponder upon this approach a bit.  
First of all, everything has a trade-off, let's list the cons and pros of multi-file approach.
Pros: 
1. smaller file is easier to browse the coherent set of tests within. You won't accidentally slip into another set of tests.
2. you may say it would be easier to locate a specific set of tests this way, but I would argue that with intelliJ it is just just as easy to locate a class as a file. (use double shift to navigate to a thing)

Cons: 
1. Naturally you want to organize test files into some hierarchical structure, for example: group together all tests against `Worker`. Using filenames isn't a very good organization because when you don't have a comprehensive view of the code base, you wouldn't be able to tell if `WorkerUnregisteringBusySpec` is a spec about `Worker` or a class called `WorkerUnregisteringBusy`. It is especially true since the word "UnregisteringBusy" isn't very clear in its meaning.  
2. in general the more files the harder to browse the code base.

So basically it can be summarized as a trade-off between reading the `Worker` tests and the browsing the code base. I think the multi-file approach would be less controversial if we mitigate the cons. Maybe instead of grouping the tests into states of Worker (which is natural only when you **know** the states), we grouping based on more clear concepts. For example,  maybe group them into a `WorkerWorkingSpec` and a `WorkerRetiringSpec`. What do you think?

nsauro commented:
as per the last comment about test organization.  Where does idle worker fit?  It's neither working nor retiring.

kailuowang commented:
so `WorkIdle`, `WorkWorking` and `WorkRetiring` ?

nsauro commented:
👊 bump 👊 tests condensed. I think that covers all feedback?

kailuowang commented:
👍
